File: src/lib/scala/Slate.Common/src/main/scala/slate/common/Result.scala
/**
<slate_header>
  author: Kishore Reddy
  url: https://github.com/kishorereddy/scala-slate
  copyright: 2015 Kishore Reddy
  license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  desc: a scala micro-framework
  usage: Please refer to license on github for more info.
</slate_header>
  */

package slate.common

import slate.common.results.{ ResultChecks}


/**
  * Container for a Success/Failure value of type T with additional values to represent
  * a string message, code, tag, error and more.
  *
  * @tparam T      : Type T
  */
sealed abstract class Result[+T] extends Product with ResultChecks
{
  val code    : Int
  val msg     : Option[String]
  val err     : Option[Throwable]
  val tag     : Option[String]
  val ref     : Option[Any]


  def success: Boolean


  def failure: Boolean = !success


  def isEmpty: Boolean


  def isDefined: Boolean = !isEmpty


  def get: T


  def getOrElse[V >: T]( default: => V ): V =
    if (isEmpty) default else this.get


  /** If the option is nonempty, return a function applied to its value,
    *  wrapped in a Some i.e. <code>Some(f(this.get))</code>.
    *  Otherwise return <code>None</code>.
    *
    *  @param  f   the function to apply
    */
  def map[B](f: T => B): Result[B] =
    if (isEmpty) this.asInstanceOf[Result[B]]
    else SuccessResult[B](f(this.get), code, msg, tag, ref)


  /** If the option is nonempty, return a function applied to its value,
    *  wrapped in a Some i.e. <code>Some(f(this.get))</code>.
    *  Otherwise return <code>defaultValue</code>.
    *
    *  @param  f   the function to apply
    */
  def mapOrElse[V](defaultValue:V, f: T => V): V =
    if (isEmpty) defaultValue
    else f(this.get)


  /** If the result is nonempty, return a function applied to its value.
    *  Otherwise return None.
    *
    *  @param  f   the function to apply
    */
  def flatMap[B](f: T => Result[B]): Result[B] =
    if (isEmpty) this.asInstanceOf[Result[B]]
    else f(this.get)


  /** If the result is nonempty, return a function applied to its value.
    *  Otherwise return None.
    *
    *  @param  f   the function to apply
    */
  def fold[B](ifEmpty: => B)(f: T => B): B =
    if (isEmpty) ifEmpty else f(this.get)


  def toOption: Option[T] = {
    if (isEmpty) None
    else Some(this.get)
  }


  def and[V >: T](other:Result[V]):Result[V] = {
    if(this.success && other.success){
      this.asInstanceOf[Result[V]]
    }
    else if(!this.success){
      this.asInstanceOf[Result[V]]
    }
    else
      other
  }

  override def statusCode = code
}

/**
  * The success branch of the Result[T]
  * @param value   : The value for a successful result/action
  * @param code    : Numeric status code
  * @param msg     : Optional string message for more information
  * @param tag     : Optional tag used for tracking purposes
  * @param ref     : Optional reference to some object ( intentional set as Any for now )
  * @param err     : Optional exception
  * @tparam T
  */
final case class SuccessResult[+T](
                                    value  : T                       ,
                                    code   : Int                     ,
                                    msg    : Option[String]    = None,
                                    tag    : Option[String]    = None,
                                    ref    : Option[Any]       = None,
                                    err    : Option[Throwable] = None
                                  ) extends Result[T] {

   def isEmpty = false


   def get = value


   def success:Boolean = true
}


/**
  * The failure branch of the Result
  * @param code    : Numeric status code
  * @param msg     : Optional string message for more information
  * @param err     : Optional exception
  * @param tag     : Optional tag used for tracking purposes
  * @param ref     : Optional reference to some object ( intentional set as Any for now )
  * @tparam T
  */
final case class FailureResult[+T](
                                    code   : Int               = 0   ,
                                    msg    : Option[String]    = None,
                                    err    : Option[Throwable] = None,
                                    tag    : Option[String]    = None,
                                    ref    : Option[Any]       = None
                                  ) extends Result[T] {
  def isEmpty:Boolean = true


  def success:Boolean = false


  def get = throw new NoSuchElementException("NoResult.get")
}


case object NoResult extends Result[Nothing] {
  val code    : Int               = 0
  val msg     : Option[String]    = None
  val err     : Option[Throwable] = None
  val tag     : Option[String]    = None
  val ref     : Option[Any]       = None

  def isEmpty = true

  def success:Boolean = false

  def get = throw new NoSuchElementException("NoResult.get")
}

File: test/resultx/src/main/kotlin/slatekit/common/Result.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package slatekit.common

import slatekit.common.results.*

/**
<slate_header>
author: Kishore Reddy
url: https://github.com/kishorereddy/scala-slate
copyright: 2015 Kishore Reddy
license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
desc: a scala micro-framework
usage: Please refer to license on github for more info.
</slate_header>
 */


/**
 * Container for a Success/Failure value of type T with additional values to represent
 * a string message, code, tag, error and more.
 *
 * @tparam T      : Type T
 */
@Suppress("UNCHECKED_CAST")
sealed class Result<out E, out T> {
    abstract val success: Boolean
    abstract val code: Int
    abstract val msg: String


    companion object {


        inline fun <T> of(f: () -> T): Result<String, T> =
            try {
                Success(f())
            } catch (e: Exception) {
                val err = e.message ?: ""
                Failure(err, FAILURE,err)
            }


        inline fun <T> attempt(f: () -> T): Result<Exception, T> =
            try {
                Success(f())
            } catch (e: Exception) {
                val err = e.message ?: ""
                Failure(e, UNEXPECTED_ERROR,err)
            }

    }
}


/**
 * Success branch of the Result
 */
data class Success<out T>(
        val data: T,
        override val code: Int = SUCCESS,
        override val msg: String = ""
) : Result<Nothing, T>() {

    override val success = true
}


/**
 * Failure branch of the result
 */
data class Failure<out E>(
        val err: E,
        override val code: Int = FAILURE,
        override val msg: String = ""
) : Result<E, Nothing>() {

    override val success = false
}


typealias ResultMsg<T> = Result<String   , T>
typealias ResultEx<T>  = Result<Exception, T>


inline fun <E, T1, T2> Result<E, T1>.map(f: (T1) -> T2): Result<E, T2> =
    when (this) {
        is Success -> Success(f(this.data), this.code, this.msg)
        is Failure -> this
    }


inline fun <E, T1, T2> Result<E, T1>.flatMap(f: (T1) -> Result<E, T2>): Result<E, T2> =
    when (this) {
        is Success -> f(this.data)
        is Failure -> this
    }


inline fun <E, R, T> Result<E, R>.fold(fl: (E) -> T, fr: (R) -> T): T =
    when (this) {
        is Success -> fr(this.data)
        is Failure -> fl(this.err)
    }


inline fun <E, T> Result<E, T>.getOrElse(f: () -> T): T =
    when (this) {
        is Success -> this.data
        is Failure -> f()
    }
